From 9c334a4f7940e68794ddfac2d44c848b63bda749 Mon Sep 17 00:00:00 2001
From: Joey Arhar <jarhar@chromium.org>
Date: Thu, 07 Nov 2024 17:11:44 +0000
Subject: [PATCH] Fix nested <option> traversals

Multiple crashes were occurring due to inconsistencies between option
traversals. OptionList was not considering nested <option>s to be
associated with the select. However, HTMLOptionElement::InsertedInto and
HTMLOptionElement::OwnerSelectElement were considering nested <option>s
to be associated with selects.

This patch makes the traversals consistent by disallowing nested
<option>s from creating associations with <select> elements.

Fixed: 376786406
Change-Id: Ia6e4cc7d028033885b452efe03687638446350ab
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5990895
Commit-Queue: Joey Arhar <jarhar@chromium.org>
Reviewed-by: David Baron <dbaron@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1379764}
---

--- a/third_party/blink/renderer/core/html/forms/html_opt_group_element.cc
+++ b/third_party/blink/renderer/core/html/forms/html_opt_group_element.cc
@@ -176,6 +176,10 @@
     // rather than doing a tree traversal here every time OwnerSelectElement is
     // called, which may be a lot.
     for (Node& ancestor : NodeTraversal::AncestorsOf(*this)) {
+      if (IsA<HTMLOptGroupElement>(ancestor) ||
+          IsA<HTMLOptionElement>(ancestor)) {
+        return nullptr;
+      }
       if (auto* select = DynamicTo<HTMLSelectElement>(ancestor)) {
         return select;
       }
--- a/third_party/blink/renderer/core/html/forms/html_option_element.cc
+++ b/third_party/blink/renderer/core/html/forms/html_option_element.cc
@@ -361,6 +361,11 @@
     // rather than doing a tree traversal here every time OwnerSelectElement is
     // called, which may be a lot.
     for (Node& ancestor : NodeTraversal::AncestorsOf(*this)) {
+      if (IsA<HTMLOptionElement>(ancestor)) {
+        // Don't associate nested <option>s with <select>s. This matches the
+        // traversals in OptionList and HTMLOptionElement::InsertedInto.
+        return nullptr;
+      }
       if (auto* select = DynamicTo<HTMLSelectElement>(ancestor)) {
         return select;
       }
@@ -494,9 +499,13 @@
   // OptionInserted. Otherwise, if this option is being inserted into a <select>
   // ancestor, then we must call OptionInserted on it.
   bool passed_insertion_point = false;
-  for (Node* ancestor = parentNode(); ancestor;
-       ancestor = ancestor->parentNode()) {
-    if (ancestor == insertion_point) {
+  for (Node& ancestor : NodeTraversal::AncestorsOf(*this)) {
+    if (IsA<HTMLOptionElement>(ancestor)) {
+      // Don't call OptionInserted() on nested <option>s. This matches the
+      // traversals in OptionList and OwnerSelectElement.
+      break;
+    }
+    if (&ancestor == &insertion_point) {
       passed_insertion_point = true;
     }
     if (auto* select = DynamicTo<HTMLSelectElement>(ancestor)) {
@@ -553,19 +562,22 @@
     return;
   }
 
-  for (Node* ancestor = parentNode(); ancestor;
-       ancestor = ancestor->parentNode()) {
+  for (Node& ancestor : NodeTraversal::AncestorsOf(*this)) {
     // If this option is still associated with a <select> inside the detached
     // subtree, then we should not call OptionRemoved() because we don't call
     // OptionInserted() in the corresponding attachment case. Also, APIs like
     // select.options should still work when the <select> is detached.
-    if (IsA<HTMLSelectElement>(ancestor)) {
+    // Nested options should not be associated with selects.
+    if (IsA<HTMLSelectElement>(ancestor) || IsA<HTMLOptionElement>(ancestor)) {
       return;
     }
   }
 
-  for (Node* ancestor = &insertion_point; ancestor;
-       ancestor = ancestor->parentNode()) {
+  for (Node& ancestor : NodeTraversal::InclusiveAncestorsOf(insertion_point)) {
+    if (IsA<HTMLOptionElement>(ancestor)) {
+      // Nested options should not be associated with selects.
+      return;
+    }
     if (auto* select = DynamicTo<HTMLSelectElement>(ancestor)) {
       SetTextOnlyRendering(true);
       select->OptionRemoved(*this);
--- a/third_party/blink/renderer/core/html/parser/html_tree_builder.cc
+++ b/third_party/blink/renderer/core/html/parser/html_tree_builder.cc
@@ -2113,8 +2113,8 @@
           }
         }
       }
-    }
       return;
+    }
     case HTMLTag::kForm:
       if (!IsParsingTemplateContents()) {
         Element* node = tree_.TakeForm();
--- a/third_party/blink/web_tests/external/wpt/html/semantics/forms/the-select-element/customizable-select/nested-options.tenative.html
+++ b/third_party/blink/web_tests/external/wpt/html/semantics/forms/the-select-element/customizable-select/nested-options.tenative.html
@@ -2,6 +2,7 @@
 <link rel=author href="mailto:jarhar@chromium.org">
 <link rel=help href="https://bugs.chromium.org/p/chromium/issues/detail?id=1422275">
 <link rel=help href="https://chromium-review.googlesource.com/c/chromium/src/+/5441435/1#message-cd8841d92a672a0276ab536dfaf3a20e93d5e6e3">
+<link rel=help href="https://issues.chromium.org/issues/376786406">
 <script src="/resources/testharness.js"></script>
 <script src="/resources/testharnessreport.js"></script>
 <script src="/resources/testdriver.js"></script>
@@ -34,10 +35,9 @@
 o1.appendChild(o2);
 
 test(() => {
-  assert_equals(select.options.length, 2, 'select.options.length');
+  assert_equals(select.options.length, 1, 'select.options.length');
   assert_equals(select.options[0], o1, 'select.options[0]');
-  assert_equals(select.options[1], o2, 'select.options[1]');
-}, 'Nested <options> should be listed in <select> IDLs.');
+}, 'Nested <options> not should be listed in <select> IDLs.');
 
 promise_test(async () => {
   await test_driver.bless();
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/semantics/forms/the-select-element/customizable-select/selected-option-parser-crash.html
@@ -0,0 +1,8 @@
+<!DOCTYPE html>
+<select disabled>
+  <option id=one selected>one</option>
+  <option id=two disabled>two
+    <table><div><div</table>
+  </option>
+  <option id=three selected>three</option>
+</select>
--- a/third_party/blink/web_tests/virtual/customizable-select-disabled/external/wpt/html/semantics/forms/the-select-element/customizable-select/nested-options.tenative-expected.txt
+++ b/third_party/blink/web_tests/virtual/customizable-select-disabled/external/wpt/html/semantics/forms/the-select-element/customizable-select/nested-options.tenative-expected.txt
@@ -1,7 +1,5 @@
 This is a testharness.js-based test.
 [FAIL] The HTML parser should disallow nested options in select datalist.
   assert_equals: expected "\\n  <datalist>\\n    <option id=\\"o1\\">\\n      parent\\n      </option><option id=\\"o2\\">child</option>\\n\\n  </datalist>\\n" but got "\\n  \\n    <option id=\\"o1\\">\\n      parent\\n      </option><option id=\\"o2\\">child</option>\\n\\n  \\n"
-[FAIL] Nested <options> should be listed in <select> IDLs.
-  assert_equals: select.options.length expected 2 but got 1
 Harness: the test ran to completion.
 
